### PR TITLE
Cache IndexSchema on SchemaSimilarityFactory to avoid NPE with shared schemas

### PR DESCRIPTION
**Pull request by johnmbw**
Using `<bool name="shareSchema">true</bool>` means that a single `SchemaSimilarityFactory` would be shared between multiple cores. Therefore `inform(SolrCore)` would be called multiple times.  This leads to a race condition when the `Similarity` object is being used for query at the same time as a new core is being loaded.  This results in a `NullPointerException` as the new `SolrCore` is not yet fully initialised.

Built on top of https://github.com/BrandwatchLtd/bw-lucene-solr/pull/55

### CODE REVIEW

**Comment by timatbw** [comment]
Oh there's a bunch of other commits, right.. will look at them separately

**Comment by johnmbw** [comment]
yeah that's to fix the tests mostly

**Review by spmason — APPROVED** [review]
LGTM, would this be something that would be portable to solr 9?

**Comment by johnmbw** [comment]
Yeah the fix itself should be I think - that class doesn't seem to have changed much.

File: solr/core/src/java/org/apache/solr/search/similarities/SchemaSimilarityFactory.java [hunk 1]
```diff
@@ -153,6 +156,17 @@ public Similarity get(String name) {
       }
     }
 
+    private IndexSchema getSchema() {
+      // if we're using shareSchema in our config we might actually have an uninitialised
```
Comment (timatbw) [inline solr/core/src/java/org/apache/solr/search/similarities/SchemaSimilarityFactory.java:160]: I did wonder if another approach would be to only do a one-shot initialisation in `inform` i.e. if the core is already not null then basically do nothing. But I don't know enough about this code to know if genuine replacement calls can occur later and need to replace the core

Comment (johnmbw) [reply solr/core/src/java/org/apache/solr/search/similarities/SchemaSimilarityFactory.java:160]: Sadly that doesn't work in `inform` as that get's called during the `SolrCore` constructor, before the index has actually been initialised.  There should probably just be some separate method to just set the schema, but yeah it's knowing how the rest of the code works that makes it tricky.  Plus I wasn't quite sure about how schema reloads would work here too.

**Review by timatbw — APPROVED** [review]
LGTM.. we'll want these in bw 8.11.2 as well

**Pull request merged**